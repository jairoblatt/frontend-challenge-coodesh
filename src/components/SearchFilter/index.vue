<template>
  <div v-clickOutside="clickOutsideHandle" class="search__filter">
    <span class="search__title">
      <slot />
    </span>

    <div @click="showOptions = !showOptions" class="filter__input">
      <span> {{ resolveName }} </span>
      <Btn> <Icon :icon="openedIcon" /> </Btn>
    </div>

    <transition name="fade-top-2x">
      <ul v-show="showOptions" class="filter__options">
        <li
          v-for="(option, key) in options"
          :key="`option-${key}`"
          :class="{ 'option--active': option.value === value }"
          @click="selectOption(option)"
        >
          {{ option.name }}
        </li>
      </ul>
    </transition>
  </div>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";
import { clickOutside } from "@/directives";

interface IOption {
  value: number | string;
  name: string;
}

export default Vue.extend({
  props: {
    options: Array as PropType<IOption[]>,
    value: [String, Number] as PropType<string | number>,
  },

  directives: { clickOutside },

  data: () => ({
    showOptions: false,
  }),

  computed: {
    openedIcon(): string {
      return this.showOptions ? "mdi-chevron-up" : "mdi-chevron-down";
    },

    resolveName(): string {
      const { options, value } = this;
      let optionsSelected: IOption | undefined;

      if (Array.isArray(options) && value) {
        optionsSelected = options.find((option: IOption) => {
          return option.value === value;
        });
      }

      return optionsSelected?.name || "";
    },
  },

  methods: {
    selectOption(option: IOption) {
      if (this.value !== option.value) {
        this.$emit("input", option.value);
        this.showOptions = false;
      }
    },

    clickOutsideHandle() {
      this.showOptions = false;
    },
  },
});
</script>

<style lang="postcss" scoped>
.search__filter {
  @apply relative ml-2 h-full w-full;
}

@screen md {
  .search__filter {
    width: 12rem;
  }
}

.search__title {
  @apply text-light-onSurfacePrimary font-semibold text-base;
}

.filter__input {
  @apply bg-white flex cursor-pointer items-center justify-between rounded-lg mt-1 border border-gray-100;
}

.filter__input span {
  @apply outline-none text-light-onSurfacePrimary pl-3 whitespace-nowrap overflow-ellipsis;
}

.filter__input >>> i {
  @apply text-light-onSurfacePrimary;
}
.filter__options {
  @apply absolute cursor-pointer z-30 rounded-sm mt-1 border-gray-100 p-1 bg-white shadow-lg text-light-onSurfacePrimary w-full overflow-y-auto;

  max-height: 15rem;
}

.filter__options li {
  @apply p-2 border border-gray-100 hover:bg-gray-50  mt-1 rounded-lg transition-colors;
}

.option--active {
  @apply bg-gray-100;
}
</style>
