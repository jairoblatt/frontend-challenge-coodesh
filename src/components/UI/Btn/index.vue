<template>
  <button v-on="$listeners" :class="classes">
    <spinner v-show="loading" size="21" />
    <slot v-show="!loading" />
  </button>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";

export default Vue.extend({
  name: "Btn",

  props: {
    primary: Boolean as PropType<boolean>,
    secondary: Boolean as PropType<boolean>,
    block: Boolean as PropType<boolean>,
    loading: Boolean as PropType<boolean>,
    small: Boolean as PropType<boolean>,
    outlined: Boolean as PropType<boolean>,
  },

  computed: {
    classes() {
      return {
        btn: true,
        "btn--primary": this.primary,
        "btn--secondary": this.secondary,
        "btn--block": this.block,
        "btn--small": this.small,
        "btn--outlined": this.outlined,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.btn {
  @apply rounded-md px-5 py-3 text-sm outline-none font-semibold text-white transition-colors;
}

.btn--primary {
  @apply bg-primary-base hover:bg-primary-light;
}

.btn--secondary {
  @apply bg-secondary-gray;
}

.btn--block {
  @apply w-full;
}

.btn:disabled {
  @apply bg-gray-300 text-gray-900 cursor-auto;
}

.btn--small {
  @apply px-7 py-2;
}

.btn--outlined {
  @apply border border-gray-100 text-light-onSurfacePrimary hover:border-black hover:text-black;
}
</style>
