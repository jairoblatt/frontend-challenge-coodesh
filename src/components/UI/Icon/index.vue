<template>
  <i v-if="iconSanitize" :class="`mdi ${iconSanitize}`" :style="styles"></i>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Icon",

  props: {
    size: {
      type: [String, Number],
      default: 14,
    } as PropOptions<string | number>,

    color: String as PropType<string>,

    icon: {
      type: String,
      default: "",
    } as PropOptions<string>,
  },

  computed: {
    iconSanitize(): string {
      return this.$slots.default?.[0]?.text || this.icon;
    },

    styles(): CSS.Properties {
      return {
        fontSize: this.size + "px",
        ...(this.color && { color: this.color }),
      };
    },
  },
});
</script>
