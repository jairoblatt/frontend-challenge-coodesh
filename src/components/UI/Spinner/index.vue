<template>
  <div class="spinner-container">
    <span class="spinner" :style="styles"></span>
  </div>
</template>
<script lang="ts">
import Vue, { PropOptions } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Spinner",

  props: {
    size: {
      type: [String, Number],
      default: "inherit",
    } as PropOptions<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        minHeight: `${this.size}px`,
        minWidth: `${this.size}px`,
      };
    },
  },
});
</script>
<style lang="postcss" scoped>
.spinner-container {
  @apply relative flex justify-center items-center min-h-full min-w-full;
}

.spinner {
  @apply pointer-events-none border-solid rounded-full  border-primary-base  p-1;

  border-width: 0.3rem;
  border-top-color: white;
  animation: spinner 0.5s linear infinite;
}

@keyframes spinner {
  100% {
    transform: rotate(360deg);
  }
}
</style>
