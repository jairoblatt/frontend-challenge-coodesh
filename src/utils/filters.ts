export const genders = [
  {
    value: "",
    name: "All",
  },
  {
    value: "female",
    name: "Female",
  },
  {
    value: "male",
    name: "Male",
  },
];

export const nationalities = [
  {
    value: "",
    name: "All",
  },
  {
    value: "au",
    name: "Australian",
  },
  {
    value: "br",
    name: "Brazilian",
  },
  {
    value: "ca",
    name: "Canadian",
  },
  {
    value: "ch",
    name: "Switzerland",
  },
  {
    value: "de",
    name: "German",
  },
  {
    value: "dk",
    name: "Danish",
  },
  {
    value: "es",
    name: "Spanish",
  },
  {
    value: "fi",
    name: "Finnish",
  },
  {
    value: "fr",
    name: "French",
  },
  {
    value: "gb",
    name: "United Kingdom",
  },
  {
    value: "ie",
    name: "Irish",
  },
  {
    value: "ir",
    name: "Iraqi",
  },
  {
    value: "no",
    name: "Norwegian",
  },
  {
    value: "nl",
    name: "Dutch",
  },
  {
    value: "nz",
    name: "New Zealand",
  },
  {
    value: "tr",
    name: "Turkey",
  },
  {
    value: "us",
    name: "American",
  },
];
