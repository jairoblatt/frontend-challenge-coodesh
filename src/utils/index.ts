export * from "./sleep";
export * from "./filters";
