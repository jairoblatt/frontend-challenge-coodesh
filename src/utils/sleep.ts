export function sleep(milliseconds?: number) {
  return new Promise<void>((resolve) => {
    setTimeout((): void => resolve(), milliseconds || 1);
  });
}
