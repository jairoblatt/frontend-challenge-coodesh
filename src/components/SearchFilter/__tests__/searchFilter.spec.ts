import { mount, Wrapper, MountOptions, createLocalVue } from "@vue/test-utils";
import { transitionStub } from "@/utils/tests";
import SearchFilter from "@/components/SearchFilter/index.vue";
import Icon from "@/components/UI/Icon/index.vue";
import Btn from "@/components/UI/Btn/index.vue";
import Spinner from "@/components/UI/Spinner/index.vue";

const localVue = createLocalVue();
localVue.component("Icon", Icon);
localVue.component("Btn", Btn);
localVue.component("Spinner", Spinner);

type Instance = InstanceType<typeof SearchFilter>;
let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

const optionsMock = [
  {
    value: 1,
    name: "one",
  },
  {
    value: 2,
    name: "two",
  },
  {
    value: 3,
    name: "three",
  },
];

describe("SearchFilter.vue ", () => {
  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(SearchFilter, {
        ...options,
        localVue,
        stubs: {
          Transition: transitionStub(),
        },
      });
    };
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should open the options menu", () => {
    const wrapper = mountFunction();

    wrapper.find(".filter__input").trigger("click");
    expect((wrapper.vm as any).showOptions).toBe(true);
  });

  it("Should resolve and display the option name when the options is selected", () => {
    const wrapper = mountFunction({
      propsData: {
        options: optionsMock,
        value: optionsMock[0].value,
      },
    });
    const selectedOptionElement = wrapper.find(".filter__input").find("span");
    expect(selectedOptionElement.text()).toBe("one");
  });

  it("Should display the current icon when the options is closed or opened", async () => {
    const wrapper = mountFunction();

    const iconComponent = wrapper.find("i");
    expect(iconComponent.classes()).toEqual(["mdi", "mdi-chevron-down"]);

    wrapper.setData({ showOptions: true });
    await wrapper.vm.$nextTick();
    expect(iconComponent.classes()).toEqual(["mdi", "mdi-chevron-up"]);
  });

  it("Should render the default slot content when him is passed", () => {
    const slot = `<h5> Hi </h5>`;

    const wrapper = mountFunction({
      slots: {
        default: slot,
      },
    });

    const slotElement = wrapper.find("h5").exists();
    expect(slotElement).toBe(true);
  });

  it('Should select option when the "selectOption" is called only if the value !== of the option selected', () => {
    const wrapper = mountFunction({
      propsData: {
        options: optionsMock,
        value: optionsMock[2].value,
      },
    });

    const firstOptions = wrapper
      .find(".filter__options")
      .findAll("li")
      .at(0);

    firstOptions.trigger("click");

    expect(wrapper.emitted("input")).toBeTruthy();
    expect(wrapper.emitted("input")?.length).toBe(1);
    expect(wrapper.emitted("input")).toEqual([[optionsMock[0].value]]);
  });

  it('Should apply the css class "option--active" when the option is selected', async () => {
    const wrapper = mountFunction({
      propsData: {
        options: optionsMock,
        value: optionsMock[0].value,
      },
    });

    const optionActiveElement = wrapper.find(".option--active").exists();
    expect(optionActiveElement).toBe(true);
  });

  it("Should display 3 options", () => {
    const wrapper = mountFunction({
      propsData: {
        options: optionsMock,
      },
    });
    const optionsItemsElement = wrapper.find(".filter__options").findAll("li");
    expect(optionsItemsElement).toHaveLength(3);
  });
});
