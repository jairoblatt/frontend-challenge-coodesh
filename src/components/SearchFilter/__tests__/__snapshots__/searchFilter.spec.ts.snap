// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFilter.vue  Should mount the component 1`] = `
"<div class=\\"search__filter\\"><span class=\\"search__title\\"></span>
  <div class=\\"filter__input\\"><span>  </span> <button class=\\"btn\\">
      <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div> <i class=\\"mdi mdi-chevron-down\\" style=\\"font-size: 14px;\\"></i>
    </button></div>
  <transition-stub name=\\"fade-top-2x\\">
    <ul class=\\"filter__options\\" style=\\"display: none;\\"></ul>
  </transition-stub>
</div>"
`;
